### PR TITLE
feat: add PyTorch DDP beginner guide for HyperPod (Slurm + EKS)

### LINKED ISSUE #133: [Content] Add PyTorch DDP Beginner Guide for SageMaker HyperPod
## Summary

Proposing a new workshop/guide under `03-workshops/` that teaches PyTorch Distributed Data Parallel (DDP) from scratch, targeting Data Scientists and Data Engineers who know single-GPU PyTorch but haven't explored multi-node training.

## Content Overview

The guide covers:
1. **Introduction** — What DDP is, when to use it (and when NOT to), DDP vs DataParallel
2. **Network Collectives** — AllReduce, Broadcast, AllGather explained with diagrams
3. **DataLoaders in DDP** — DistributedSampler, effective batch size equation, performance tips
4. **Step-by-step DDP script** — Evolves from single-GPU baseline to full multi-node script using Qwen2.5-0.5B
5. **HyperPod + Slurm** — sbatch script, torchrun integration, EFA/NCCL config, auto-resume
6. **HyperPod + EKS** — Dockerfile, PyTorchJob manifest, SIGTERM handling, auto-resume
7. **Distributed Checkpointing** — Atomic saves, resume logic, torch.distributed.checkpoint (DCP)
8. **What's Next** — Gradient accumulation with no_sync, mixed precision, torch.compile, FSDP

## Key Features
- Progressive "evolve the code" pedagogy — each step adds one concept
- Complete, runnable code examples (not pseudocode)
- Both Slurm AND EKS orchestration paths
- HyperPod-specific content (SIGTERM handling, auto-resume, EFA config)
- 4 SVG diagrams (AllReduce, architecture, DataLoader flow, checkpointing)
- Troubleshooting tables for common multi-node issues

## Target Audience
Data Scientists and Data Engineers familiar with PyTorch who want to scale to multiple GPUs/nodes on SageMaker HyperPod.

## Proposed Location
`website/docs/03-workshops/01-distributed-training-beginner/`

### PR DESCRIPTION
**Pull request by paragao**
## What

Adds a 7-page workshop guide teaching PyTorch Distributed Data Parallel (DDP) for multi-node pre-training on SageMaker HyperPod. Covers both Slurm and EKS orchestration.

Closes #133

## Content Structure

```
website/docs/03-workshops/01-distributed-training-beginner/
├── _category_.json
├── 00-introduction.mdx        (Intro + Network Collectives)
├── 01-dataloaders.mdx         (DataLoaders + DistributedSampler)
├── 02-ddp-script.mdx          (Step-by-step DDP script with Qwen2.5-0.5B)
├── 03-slurm.mdx               (HyperPod + Slurm orchestration)
├── 04-eks.mdx                 (HyperPod + EKS orchestration)
├── 05-checkpointing.mdx       (Distributed checkpointing + DCP)
├── 06-whats-next.mdx          (Advanced topics: FSDP, mixed precision, torch.compile)
└── assets/
    ├── diagram_allreduce.svg
    ├── diagram_ddp_architecture.svg
    ├── diagram_dataloader.svg
    └── diagram_checkpointing.svg
```

## Highlights

- **Progressive pedagogy** — evolves from single-GPU baseline to full multi-node DDP step by step
- **Both orchestrators** — Slurm (sbatch + torchrun) and EKS (PyTorchJob CRD)
- **HyperPod-native** — SIGTERM handling, auto-resume, EFA/NCCL configuration
- **Production patterns** — gradient clipping, LR scaling, `model.no_sync()`, proper metric aggregation via `dist.all_reduce`, distributed checkpointing
- **Dark mode compatible** — SVG diagrams wrapped for visibility in dark theme
- **Model:** Qwen2.5-0.5B (open source, small enough to demo, realistic enough to learn from)

## Target Audience

Data Scientists and Data Engineers familiar with PyTorch who want to scale to multiple GPUs/nodes on SageMaker HyperPod.

## Testing

- [x] `npm run build` passes cleanly
- [x] All SVG paths resolve correctly
- [x] MDX compiles without errors
- [x] Navbar dropdown entry links correctly

### CODE REVIEW

**Review by yhou-uk — COMMENTED** [review]
Really nice work on this guide — it's going to save a lot of people their first painful week with multi-node DDP. A few things I think it does especially well:

- **The mental model up front.** Leading file 00 with `DDP = Full Model Replication + Data Sharding + AllReduce Gradient Sync` gives readers an anchor before any code, and the DP-vs-DDP table makes the "always prefer DDP" takeaway land immediately.
- **It teaches the non-obvious stuff that actually bites people** — `sampler.set_epoch()`, `no_sync()` for accumulation, `model.module` for checkpoints, `find_unused_parameters`, effective-batch-size math, and LR scaling. These are exactly the gotchas beginners hit, and they're explained at the right depth.
- **The progression is well sequenced** — single-GPU baseline → process group → DDP wrap → sampler → full script — so the reader builds the script up instead of being handed a wall of code.
- **The troubleshooting tables** at the end of the Slurm/EKS chapters are genuinely useful and map symptoms → cause → fix the way someone debugging at 2am would want.

I left a few specific comments inline — a couple are correctness issues worth fixing before merge, the rest are accuracy/consistency tweaks.

File: website/docs/03-workshops/01-distributed-training-beginner/02-ddp-script.mdx [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+            
+            total_loss += loss.item() * ACCUMULATION_STEPS
+            num_steps += 1
+            
+            # ─── Logging with proper metric aggregation ───────────────────
+            if step % 100 == 0 and not is_accumulation_step:
+                # Aggregate loss across all ranks for accurate global metric
+                loss_tensor = torch.tensor([loss.item() * ACCUMULATION_STEPS], device=local_rank)
+                dist.all_reduce(loss_tensor, op=dist.ReduceOp.AVG)
+                log(f"  Epoch {epoch} | Step {step} | Global Avg Loss: {loss_tensor.item():.4f}")
```
Comment (yhou-uk) [inline website/docs/03-workshops/01-distributed-training-beginner/02-ddp-script.mdx]: **Per-step loss logging never fires with the default config.**

This guard needs two conditions that can't both be true with the shipped settings:

```python
if step % 100 == 0 and not is_accumulation_step:
```

With `ACCUMULATION_STEPS = 2`, a sync step (`not is_accumulation_step`) only happens on **odd** `step` values, while `step % 100 == 0` is only ever true on **even** values — the two sets never intersect, so this block (and its `dist.all_reduce` global-loss print) runs **zero times**. The epoch-end summary on L392–395 still prints, so training looks fine; you just never get per-step progress.

**Suggested fix** — drive logging off an optimizer-step counter instead of the accumulation parity (needs `optim_step = 0` before the loop, so leaving this as reference rather than a one-click suggestion):

```python
if not is_accumulation_step:
    torch.nn.utils.clip_grad_norm_(model.parameters(), MAX_GRAD_NORM)
    optimizer.step()
    optimizer.zero_grad()
    optim_step += 1

    if optim_step % 50 == 0:   # log every 50 optimizer updates
        loss_tensor = torch.tensor([loss.item() * ACCUMULATION_STEPS], device=local_rank)
        dist.all_reduce(loss_tensor, op=dist.ReduceOp.AVG)
        log(f"  Epoch {epoch} | Optim step {optim_step} | Global Avg Loss: {loss_tensor.item():.4f}")
```

File: website/docs/03-workshops/01-distributed-training-beginner/05-checkpointing.mdx [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+# Distributed Checkpointing: Save and Resume
+
+<div style={{background: 'white', padding: '1rem', borderRadius: '8px', marginBottom: '1rem'}}>
+
+![Checkpointing Flow](./assets/diagram_checkpointing.png)
+
+</div>
+
+*Figure 4: Rank 0 saves to shared filesystem; all ranks load in parallel on resume.*
```
Comment (yhou-uk) [inline website/docs/03-workshops/01-distributed-training-beginner/05-checkpointing.mdx]: **This diagram contradicts the text right below it.**

The caption and the "Core Principle" section describe the **rank-0 single-file** approach ("you only need one rank to save… rank 0 writes"). But `diagram_checkpointing.png` actually shows the **DCP sharded** flow — all four ranks each writing `model_shard_N + optim_state_N.pt`, with the resume side labeled *"DCP reshards"* / *"torch.distributed.checkpoint (DCP)."* That's the opposite of the caption, and DCP isn't introduced until the optional subsection further down — so a beginner sees a 4-rank sharded picture next to "rank 0 saves."

**Suggested fix (lightest):** move this image down under the **"Using `torch.distributed.checkpoint` (DCP)"** subsection where it's accurate, and reword the caption to something like *"With DCP, every rank writes its own shard in parallel and the checkpoint is resharded on resume."* For the top section, either add a simple rank-0 → shared-storage graphic or just let the text stand.

File: website/docs/03-workshops/01-distributed-training-beginner/02-ddp-script.mdx [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+    torch.cuda.manual_seed_all(42)
+    
+    return rank, local_rank, world_size
+
+def cleanup():
+    """Destroy the process group."""
+    dist.destroy_process_group()
+```
+
+> **NCCL Timeout:** The default timeout is 30 minutes. For large clusters (32+ nodes) or when debugging, increase it. If a rank hangs longer than the timeout, PyTorch raises an error instead of hanging indefinitely.
```
Comment (yhou-uk) [inline website/docs/03-workshops/01-distributed-training-beginner/02-ddp-script.mdx]: **NCCL default timeout is 10 minutes, not 30.**

For the NCCL backend in PyTorch 2.x the default is **10 minutes** — `kProcessGroupNCCLDefaultTimeout = std::chrono::milliseconds(10 * 60 * 1000)` ([PyTorch source](https://github.com/pytorch/pytorch/blob/main/torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp)). (30 min was the old pre-split default for all backends — easy mix-up.)

```suggestion
> **NCCL Timeout:** The default timeout is 10 minutes. For large clusters (32+ nodes) or when debugging, increase it. If a rank hangs longer than the timeout, PyTorch raises an error instead of hanging indefinitely.
```

Two related spots to keep consistent: the script sets `timeout=timedelta(minutes=10)` (L203) commented *"Increase for large clusters"* — but 10 min **is** the default, so consider bumping the example to `minutes=30` so it's a genuine increase; and the `03-slurm.mdx` troubleshooting row ("Timeout after 30min") should reference the 10-minute default too.

File: website/docs/03-workshops/01-distributed-training-beginner/04-eks.mdx [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+                - /workspace/train_ddp.py
+              resources:
+                requests:
+                  nvidia.com/gpu: 8
+                  memory: "512Gi"
+                  cpu: "96"
+                limits:
+                  nvidia.com/gpu: 8
+                  memory: "512Gi"
+                  cpu: "96"
```
Comment (yhou-uk) [inline website/docs/03-workshops/01-distributed-training-beginner/04-eks.mdx:87]: **EFA is configured via env vars but never actually requested, so pods won't get it.**

The pod sets `FI_PROVIDER=efa` and the troubleshooting section references `aws-efa-k8s-device-plugin`, but `resources` only ask for `nvidia.com/gpu`. Without an explicit `vpc.amazonaws.com/efa` request the scheduler won't attach EFA interfaces, so NCCL silently falls back to TCP and the multi-node speedup this chapter is about doesn't materialize. AWS's own [EKS + EFA reference manifest](https://docs.aws.amazon.com/eks/latest/userguide/node-efa.html) requests `vpc.amazonaws.com/efa` and mounts `/dev/shm`.

```suggestion
              resources:
                requests:
                  nvidia.com/gpu: 8
                  vpc.amazonaws.com/efa: 32   # request EFA interfaces (per AWS EKS EFA guide)
                  memory: "512Gi"
                  cpu: "96"
                limits:
                  nvidia.com/gpu: 8
                  vpc.amazonaws.com/efa: 32
                  memory: "512Gi"
                  cpu: "96"
```

Also worth adding a shared-memory volume (the default 64Mi `/dev/shm` commonly crashes NCCL at multi-GPU scale) — non-contiguous, so reference rather than a suggestion:

```yaml
              volumeMounts:
                - name: fsx-pvc
                  mountPath: /fsx
                - name: shmem
                  mountPath: /dev/shm
          volumes:
            - name: fsx-pvc
              persistentVolumeClaim:
                claimName: fsx-lustre-pvc
            - name: shmem
              hostPath:
                path: /dev/shm
```

File: website/docs/03-workshops/01-distributed-training-beginner/03-slurm.mdx [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+3. HyperPod replaces the node and the job can be requeued.
+
+The training script above already handles SIGTERM (see the `sigterm_handler`). To enable automatic requeue on the Slurm side, add:
+
+```bash
+#SBATCH --requeue                      # Allow auto-requeue on preemption
+#SBATCH --signal=SIGTERM@120           # Send SIGTERM 120s before killing
+```
+
+The script saves a checkpoint on SIGTERM, and on restart it resumes from the latest checkpoint automatically.
```
Comment (yhou-uk) [inline website/docs/03-workshops/01-distributed-training-beginner/03-slurm.mdx]: **This shows generic Slurm requeue, not HyperPod auto-resume.**

The section is billed as HyperPod's "key differentiator," but `#SBATCH --requeue` + `--signal=SIGTERM@120` is stock Slurm. Per the [HyperPod resiliency docs](https://docs.aws.amazon.com/sagemaker/latest/dg/sagemaker-hyperpod-resiliency-slurm-auto-resume.html), HyperPod auto-resume is enabled by passing **`--auto-resume=1` to `srun`** and running setup inside a single `srun` entrypoint, so the job restarts from checkpoint after the faulty node is replaced.

Related: that doc explicitly warns **not** to rely on `$SLURM_JOB_NODELIST` for auto-resume jobs (it can be stale after a node is replaced) — but the master-address resolution on L51 does exactly that. It recommends re-deriving the list with `scontrol show jobid=$SLURM_JOBID`.

**Suggested direction:**

```bash
# Re-derive the node list (do NOT use $SLURM_JOB_NODELIST — it can be stale after node replacement)
NODE_LIST=$(scontrol show jobid=$SLURM_JOBID | awk -F= '/NodeList=/{print $2}' | grep -v Exc)
MASTER_ADDR=$(scontrol show hostname $NODE_LIST | head -n 1)

# Enable HyperPod auto-resume with --auto-resume=1 on srun
srun --auto-resume=1 torchrun \
    --nnodes=$SLURM_NNODES \
    --nproc_per_node=8 \
    --rdzv_id=$SLURM_JOB_ID \
    --rdzv_backend=c10d \
    --rdzv_endpoint=$MASTER_ADDR:29500 \
    /fsx/training/ddp-demo/train_ddp.py
```

A one-liner noting that `--requeue`/`--signal` is the *generic* Slurm fallback while `--auto-resume=1` is the HyperPod feature would make the distinction crisp.

**Review by yhou-uk — APPROVED** [review]
file patch5-auto-resume.patch need to be removed before merge.
